app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Price/Query/BaseFinalPrice.php

```diff
@@ -217,11 +217,7 @@ public function getQuery(array $dimensions, string $productType, array $entityId
         $select->where("e.type_id = ?", $productType);
 
         if ($entityIds !== null) {
-            if (count($entityIds) > 1) {
-                $select->where(sprintf('e.entity_id BETWEEN %s AND %s', min($entityIds), max($entityIds)));
-            } else {
-                $select->where('e.entity_id = ?', $entityIds);
-            }
+            $select->where('e.entity_id IN (?)', $entityIds);
```

Comment:
I believe that from the performance point of view better to use two where condition, to allow MySQL use range optimization for primary index scan:
like
```suggestion
             $select->where(sprintf('e.entity_id BETWEEN %s AND %s', min($entityIds), max($entityIds)));
             $select->where('e.entity_id IN (?)', $entityIds);
```
and return case for exact one product

Reply:
@kandy wondering why MySQL cannot do such kind of optimization by itself...

Reply:
@orlangur, no, because it not common case and  base on specific usage in Magento Application

Reply:
I see, we know something about underlying data - like, range is not too big.

Still some heuristics could be applied on MySQL engine side.

Reply:
@kandy `e.entity_id IN (?)` I have serious doubts MySQL is not optimizing it well in case of only one ID.

Reply:
@orlangur - maybe because of the limit of query size:
https://stackoverflow.com/questions/4275640/mysql-in-condition-limit/4275704

It was hell to find the problem.

Reply:
Sure, there can be limits with SQL, MySQL. config parameters etc... but this is what code is there for, to do intelligent things even with possible limitations somewhere else...

Reply:
@osakoova I don't see `e.entity_id IN (?)` part in your query.